lib/Sema/BuilderTransform.cpp

```diff
@@ -1884,3 +1886,107 @@ bool TypeChecker::typeSupportsBuilderOp(
   return foundMatch;
 }
 
+Type swift::inferFunctionBuilderComponentType(NominalTypeDecl *builder) {
+  Type componentType;
+
+  SmallVector<ValueDecl *, 4> potentialMatches;
+  ASTContext &ctx = builder->getASTContext();
+  bool supportsBuildBlock = TypeChecker::typeSupportsBuilderOp(
+      builder->getDeclaredInterfaceType(), builder, ctx.Id_buildBlock,
+      /*argLabels=*/{}, &potentialMatches);
+  if (supportsBuildBlock) {
+    for (auto decl : potentialMatches) {
+      auto func = dyn_cast<FuncDecl>(decl);
+      if (!func || !func->isStatic())
+        continue;
+
+      // If we haven't seen a component type before, gather it.
+      if (!componentType) {
+        componentType = func->getResultInterfaceType();
+        continue;
+      }
+
+      // If there are inconsistent component types, bail out.
+      if (!componentType->isEqual(func->getResultInterfaceType())) {
+        componentType = Type();
+        break;
+      }
+    }
+  }
+
+  return componentType;
+}
+
+void swift::printFunctionBuilderBuildFunction(
```

Comment:
It would be great to use this for the "missing buildBlock" case too in `visitFunctionBuilderAttr`!

Reply:
Absolutely! Done in https://github.com/apple/swift/pull/33972/commits/3b8656dc6fd94a91d2c1dd3e1b98343a34d0b731, along with `buildLimitedAvailability(_:)`.